### PR TITLE
add phpstan static analysis tool

### PR DESCRIPTION
**Pull request by prondubuisi**
Fixes issues #196 

Hello @dkarlovi @bobstrecansky any input on this?

### CODE REVIEW

**Comment by prondubuisi** [comment]
Also this is the error by PHPStan at the moment

Note: Using configuration file /gitrepos/open-php/phpstan.neon.
 78/78 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------- 
  Line   Context/Context.php                                    
 ------ ------------------------------------------------------- 
  61     Unsafe usage of new static().                          
         💡 Consider making the class or the constructor final.  
  85     Unsafe usage of new static().                          
         💡 Consider making the class or the constructor final.  
  88     Unsafe usage of new static().                          
         💡 Consider making the class or the constructor final.  
  141    Unsafe usage of new static().                          
         💡 Consider making the class or the constructor final.  
 ------ ------------------------------------------------------- 

                                                                                                                   
 [ERROR] Found 4 errors                                                                                            
                                                                                                                   

Ndubuisis-MacBook-Pro:open-php prondubuisi$

File: phpstan.neon [hunk 1]
```diff
@@ -0,0 +1,10 @@
+parameters:
+    level: 0
+    paths:
+        -   Context
+        -   api
+        -   contrib
+        -   sdk
+
```
Comment (bobstrecansky) [inline phpstan.neon:8]: Extra whitespace?

File: phpstan.neon [hunk 2]
```diff
@@ -0,0 +1,10 @@
+parameters:
+    level: 0
```
Comment (bobstrecansky) [inline phpstan.neon:2]: I'm assuming we are going to start at `level 0` here and work our way up?  Is there a big difference in running different levels here?

**Review by bobstrecansky — COMMENTED** [review]
A couple questions, but I very much like the direction of this PR.  We should fix those psalm errors in a separate PR.

File: phpstan.neon (hunk 2 above)
Comment (dkarlovi) [reply phpstan.neon:2]: We can start at 4 since my PR brought us up to that IIRC.

**Comment by bobstrecansky** [comment]
@dkarlovi - did you run into that error messaging from psalm when you were making your PR?  I did some research and found some github issues related to:

`Unsafe usage of new static().`

https://github.com/phpstan/phpstan/issues/2963

**Comment by dkarlovi** [comment]
This is PHPStan, we already enabled Psalm.

Basically, new static is unsafe because the extending class can change your constructor (for example, make it have a mandatory param where your doesn't), making your code crash.

Since the error is correct, we can either fix the code or ignore the error in PHPStan config.

**Comment by bobstrecansky** [comment]
Oh my apologies, I did mean PHPStan, typo!
I think it warrants fixing if possible.

**Comment by prondubuisi** [comment]
Hello @bobstrecansky @dkarlovi  Great Convo! @dkarlovi I think we are more interested in the fix. I will be pushing this up and fixing comments so far.  How do you think I can fix the current error @dkarlovi

**Comment by dkarlovi** [comment]
Easiest is to make the constructor final

File: phpstan.neon (hunk 1 above)
Comment (prondubuisi) [reply phpstan.neon]: Fixed

File: phpstan.neon (hunk 2 above)
Comment (prondubuisi) [reply phpstan.neon]: Moved to 4

**Review by bobstrecansky — APPROVED** [review]
LGTM.

**Review by dkarlovi — APPROVED** [review]
LGTM!

**Comment by dkarlovi** [comment]
@prondubuisi would you be interested in adding mutation testing (in a separate PR)? Since the project already has a test suite with coverage, adding mutation should be pretty straightforward.

@bobstrecansky WDYT?

**Comment by bobstrecansky** [comment]
Sounds like a plan to me!  I say we merge this as is and we can create a separate ticket / PR to track mutation testing.

**Comment by prondubuisi** [comment]
> @prondubuisi would you be interested in adding mutation testing (in a separate PR)? Since the project already has a test suite with coverage, adding mutation should be pretty straightforward.
> 
> @bobstrecansky WDYT?

Sure @dkarlovi  will do.

**Pull request merged**